File: internal/render/ears.go
package render

import "fmt"

// placeholder is used when a required EARS field is empty.
const placeholder = "<missing>"

// field returns the value if non-empty, or the placeholder string.
func field(s string) string {
	if s == "" {
		return placeholder
	}
	return s
}

// EARSCriterion is an interface over the Criterion fields needed for rendering.
// We use a plain struct to avoid importing the root afspec package.
type EARSCriterion struct {
	EarsPattern    string
	System         string
	Action         string
	ReturnContract *string
	Trigger        string
	Condition      string
	ErrorCondition string
	State          string
	Feature        string
}

// RenderEARS renders a single EARS criterion to its sentence form.
// Uses the six templates from spec-format.md §5.2.1.
// Empty required fields render as "<missing>".
// A null or empty return_contract omits the "AND return" clause.
func RenderEARS(c *EARSCriterion) (string, error) {
	var sentence string

	switch c.EarsPattern {
	case "ubiquitous":
		sentence = fmt.Sprintf("THE %s SHALL %s", field(c.System), field(c.Action))
	case "event_driven":
		sentence = fmt.Sprintf("WHEN %s, THE %s SHALL %s", field(c.Trigger), field(c.System), field(c.Action))
	case "complex_event":
		sentence = fmt.Sprintf("WHEN %s AND %s, THE %s SHALL %s", field(c.Trigger), field(c.Condition), field(c.System), field(c.Action))
	case "state_driven":
		sentence = fmt.Sprintf("WHILE %s, THE %s SHALL %s", field(c.State), field(c.System), field(c.Action))
	case "unwanted":
		sentence = fmt.Sprintf("IF %s, THEN THE %s SHALL %s", field(c.ErrorCondition), field(c.System), field(c.Action))
	case "optional":
		sentence = fmt.Sprintf("WHERE %s, THE %s SHALL %s", field(c.Feature), field(c.System), field(c.Action))
	default:
		return "", fmt.Errorf("unknown EARS pattern: %q", c.EarsPattern)
	}

	// Append return contract clause if non-null and non-empty.
	if c.ReturnContract != nil && *c.ReturnContract != "" {
		sentence += fmt.Sprintf(" AND return %s", *c.ReturnContract)
	}

	return sentence, nil
}

File: packages/afspec/afspec/ears.py
"""EARS sentence rendering from criterion fields.

Renders EARS (Easy Approach to Requirements Syntax) sentences from
decomposed criterion fields using the six templates defined in
docs/spec-format.md section 6.2.1.
"""

from __future__ import annotations

from afspec.models import Criterion, EARSPattern


def render_ears_sentence(c: Criterion) -> str:
    """Render an EARS sentence from a criterion's fields.

    Uses one of six templates based on the criterion's ears_pattern:
      - ubiquitous:    THE {system} SHALL {action}
      - event_driven:  WHEN {trigger}, THE {system} SHALL {action}
      - complex_event: WHEN {trigger} AND {condition}, THE {system} SHALL {action}
      - state_driven:  WHILE {state}, THE {system} SHALL {action}
      - unwanted:      IF {error_condition}, THEN THE {system} SHALL {action}
      - optional:      WHERE {feature}, THE {system} SHALL {action}

    When return_contract is non-null, appends: " AND return {return_contract}".
    """
    pattern = c.ears_pattern

    if pattern == EARSPattern.UBIQUITOUS:
        sentence = f"THE {c.system} SHALL {c.action}"
    elif pattern == EARSPattern.EVENT_DRIVEN:
        sentence = f"WHEN {c.trigger}, THE {c.system} SHALL {c.action}"
    elif pattern == EARSPattern.COMPLEX_EVENT:
        sentence = f"WHEN {c.trigger} AND {c.condition}, THE {c.system} SHALL {c.action}"
    elif pattern == EARSPattern.STATE_DRIVEN:
        sentence = f"WHILE {c.state}, THE {c.system} SHALL {c.action}"
    elif pattern == EARSPattern.UNWANTED:
        sentence = f"IF {c.error_condition}, THEN THE {c.system} SHALL {c.action}"
    elif pattern == EARSPattern.OPTIONAL:
        sentence = f"WHERE {c.feature}, THE {c.system} SHALL {c.action}"
    else:
        msg = f"Unknown EARS pattern: {pattern}"
        raise ValueError(msg)

    if c.return_contract is not None:
        sentence += f" AND return {c.return_contract}"

    return sentence
